### DISCUSSION TITLE
Building libplacebo on Windows?

### DISCUSSION THREAD

**Discussion by hrxn** [haasn/libplacebo#231] [Q&A] [answered]
It seems like it should be supported, according to the meson build file
```
if host_machine.system() == 'windows'
```

But I couldn't find any guide/list of requirements on the project homepage or the repo.

What would be necessary to achieve this?
I want to test [plplay](https://code.videolan.org/videolan/libplacebo/-/blob/master/demos/plplay.c?ref_type=heads) with its interactive output modifications.

**Comment by haasn** [comment]
To be honest, I don't know. Maybe @kasper93 or @Nevcairiel can help?

**Comment by kasper93** [comment, marked as answer]
There is nothing special about building libplacebo on Windows.
```
meson setup build
meson compile -C build
```
works. 

Of course you need build environment with compiler and dependencies (ffmpeg for plplay). You can use MSYS2 [1] and install prebuilt ffmpeg package, see mpv guide for more details [2]. MSVC is also possible then you have to build ffmpeg yourself [3]. (and native file for meson, because tool autodetection is not that good, but when you arrive at this point I can help with that)

[1] https://www.msys2.org/wiki/MSYS2-installation/
[2] https://github.com/mpv-player/mpv/blob/master/DOCS/compile-windows.md#native-compilation-with-msys2
[3] https://trac.ffmpeg.org/wiki/CompilationGuide

With those pointer you should be able to built it, feel free to ask questions.

> list of requirements 

Look here for complete list and build steps https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-libplacebo/PKGBUILD (don't copy `-Ddemos=false` though :))

**Reply by kasper93** [reply]
I've also asked if MSYS2 guys would be interested in providing plplay and other libplacebo demo applications in their packages, so they are easily accessible for users. https://github.com/msys2/MINGW-packages/issues/19354

**Reply by hrxn** [reply]
Thanks, very helpful list of links!

What kind of ffmpeg artifacts would I need for plplay, though? 
Aren't the usual DLLs enough, i.e. ffmpeg shared builds? 
Already available from like https://github.com/BtbN/FFmpeg-Builds, for example?


List of Requirements

```
- [ ] Python (obvious and easy, has self-explanatory Windows Installers)
- [ ] Meson (distributed via pip)
- [ ] Ninja (distributed via cross-platform static binaries on their GitHub releases page, apparently)
- [ ] MSVC (Community Edition) (or are Microsoft C++ Build Tools enough? *Should* contain a standalone compiler, AFAIK)
```

Of course, then there are still libplacebo's dependencies as listed on the repo page:

- **glslang**: glslang + its related libraries (e.g. libSPIRV.so)
- **lcms**: liblcms2
- **libdovi**: libdovi
- **opengl**: glad2 (*)
- **shaderc**: libshaderc
- **vulkan**: libvulkan, python3-jinja2 (*)
- **xxhash**: libxxhash


Probably still need MinGW/MSYS2 for this, right?

**Reply by kasper93** [reply]
>  Already available from like https://github.com/BtbN/FFmpeg-Builds, for example?

Yes, those are fine.

> Probably still need MinGW/MSYS2 for this, right?

It will be infinitely easier to use MSYS2, not even MinGW alone, because you can just install prebuild packages of all dependencies and just build the libplacebo. Else you have to build dependencies manually.

**Comment by kasper93** [comment]
[libplacebo-demos](https://github.com/msys2/MINGW-packages/tree/master/mingw-w64-libplacebo-demos) package has been added for MSYS2. So now you can just install prebuilt package instead of compiling plplay yourself.

**Comment by hrxn** [comment]
Ah, nice.

Although this repo apparently only contains the `PKGBUILD` files.

Binaries are linked here, I think:
https://packages.msys2.org/base/mingw-w64-libplacebo-demos

Not sure if you'd want `clang64`, `mingw64`, `ucrt64`, although they should all work?

Seems like they are compressed with zstd, so one would also need a program able to handle these, normal 7zip won't cut it here I think.

**Reply by kasper93** [reply]
It is environment and package manager. Need to install it first, following https://www.msys2.org/wiki/MSYS2-installation and then start shell, like with clang64.exe and do `pacman -S pactoys` once and `pacboy -S libplacebo-demo:p` then you can use `plplay`

**Reply by hrxn** [reply]
I'm perfectly aware that this repo is only for the package scripts, but, according to their own [description](https://github.com/msys2/MINGW-packages#:~:text=The%20common%20way%20to%20use%20these%20packages%20are%20pre%2Dbuilt%20binary%20packages%20from%20the%20MSYS2%20MINGW64%20repo), they distribute them in binary form, so they have to be hosted somewhere.

And I'm going to try that first, to see if I can get away without installing anything 😄
